src/lib/controllers/ui-controller.js

```diff
@@ -676,72 +779,105 @@ export default class UIController {
   }
 
   showLanguageInfo (homonym) {
-    let notFound = !homonym ||
+    const notFound = !homonym ||
       !homonym.lexemes ||
       homonym.lexemes.length < 1 ||
       homonym.lexemes.filter((l) => l.isPopulated()).length < 1
-    if (this.hasUiModule('panel')) { this.getUiModule('panel').vi.showLanguageNotification(homonym, notFound) }
-    if (this.hasUiModule('popup')) { this.getUiModule('popup').vi.showLanguageNotification(homonym, notFound) }
+
+    if (notFound) {
+      let languageName
+      if (homonym) {
+        languageName = this.api.app.getLanguageName(homonym.languageID).name
+      } else if (this.store.state.app.currentLanguageName) {
+        languageName = this.store.state.app.currentLanguageName
+      } else {
+        // TODO this wil be unnecessary when the morphological adapter returns a consistent response for erors
+        languageName = this.api.l10n.getMsg('TEXT_NOTICE_LANGUAGE_UNKNOWN')
+      }
+      const message = this.api.l10n.getMsg('TEXT_NOTICE_CHANGE_LANGUAGE', { languageName: languageName })
+      this.store.commit(`ui/setNotification`, { text: message, important: true, showLanguageSwitcher: true })
+    }
   }
 
   showStatusInfo (selectionText, languageID) {
-    if (this.hasUiModule('panel')) { this.getUiModule('panel').vi.showStatusInfo(selectionText, languageID) }
-    if (this.hasUiModule('popup')) { this.getUiModule('popup').vi.showStatusInfo(selectionText, languageID) }
+    this.store.commit(`app/setStatusData`, { text: selectionText, languageID: languageID })
   }
 
+  // TODO: Do we need this function
   showErrorInfo (errorText) {
-    if (this.hasUiModule('panel')) { this.getUiModule('panel').vi.showErrorInformation(errorText) }
+    this.store.commit(`ui/setNotification`, { text: errorText, important: true })
   }
 
+  // TODO: Do we need this function
   showImportantNotification (message) {
-    if (this.hasUiModule('panel')) { this.getUiModule('panel').vi.showImportantNotification(message) }
-    if (this.hasUiModule('popup')) { this.getUiModule('popup').vi.showImportantNotification(message) }
+    this.store.commit(`ui/setNotification`, { text: message, important: true })
   }
 
   changeTab (tabName) {
     const statusAvailable = Boolean(this.api.settings.contentOptions.items.verboseMode.currentValue === 'verbose')
     // If tab is disabled, switch to a default one
     if (
-      (!this.store.state.app.tabState.hasOwnProperty(tabName)) ||
+      /* (!this.store.state.app.tabState.hasOwnProperty(tabName)) || */
```

Comment:
Yes, it is possible to switch to a non-existent tab, but I thought probably we don't need this check here. Here is my reasoning.

1. This piece of code does input data validation. `changeTab()` is called from within our code, not by any outside clients. In our code, it's more efficient to debug it and make sure we do not call an incorrect tab anywhere. This will allow us to eliminate this check and save a couple of microseconds 🙂. If it would be called by the user, we definitely should expect an incorrect value and check for that. But if it's our code, and it's tested to be working properly, it makes no sense to surround every function in our app with error checking statements.

2. The second reason is more utilitarian 🙂. In order to make this check, we have to keep a list of all valid tabs and update them once we add or remove a new tab. This makes code maintenance more complex. By eliminating this check, we can keep only the active tab name and forget about a list of all available tabs.

What do you think? Shall we keep this check or can we remove it?

Reply:
ah, the 2nd point makes me remember maybe what caused us to add this in the first place.  The inflection tab comes and goes depending upon the current query.  But if it's gone, then we should never see a UI element which allows us to change to it, I guess?

Reply:
We do have conditional checks on all tab controls that opens such panels. So yes, that will be impossible to change to that tab's panel if condition is false (e.g. no inflection data).